ImputationPipeline/CreateAggregationSets/tests/CreateAggregationSetsTest.py

```diff
@@ -0,0 +1,192 @@
+import ImputationPipeline.CreateAggregationSets.CreateAggregationSets as CreateAggregationSets
+import unittest
+import os
+import firecloud.api as fapi
+import json
+import random
+
+current_dir = os.path.dirname(__file__)
+resources_dir = os.path.join(current_dir, "resources")
+
+
+def json_lists_to_sets(o):
+    if isinstance(o, list):
+        return {json_lists_to_sets(v) for v in o}
+    elif isinstance(o, dict):
+        return {k: json_lists_to_sets(v) for k, v in o.items()}
+    return o
+
+
+def build_sample_sets_lite(sample_sets: list):
+    sample_sets_lite = dict()
+    for sample_set in sample_sets:
+        name = sample_set["name"]
+        group = sample_set["attributes"]["group"]
+        lab_batch = sample_set["attributes"]["lab_batch"]
+        samples = {sample["entityName"] for sample in sample_set["attributes"]["samples"]["items"]}
+        delivered = sample_set["attributes"]["delivered"]
+        redeliver = sample_set["attributes"]["redeliver"]
+        sample_sets_lite[name] = {"group": group, "lab_batch": lab_batch, "samples": samples, "delivered": delivered,
+                                  "redeliver": redeliver}
+    return sample_sets_lite
+
+
+class IntegrationTests(unittest.TestCase):
+    workspace_namespace = None
+    workspace_name = None
+    maxDiff = None
+
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.workspace_namespace = os.getenv("WORKSPACE_NAMESPACE")
+        if cls.workspace_namespace is None:
+            raise (RuntimeError("Environment variable WORKSPACE_NAMESPACE must be set to run tests"))
+        workspace_hash = random.getrandbits(64)
+        cls.workspace_name = f'test_workspace_{workspace_hash:016x}'
+        print(f"Creating test workspace {cls.workspace_namespace}/{cls.workspace_name}")
+        r = fapi.create_workspace(cls.workspace_namespace, cls.workspace_name)
+        fapi._check_response_code(r, 201)
+
+    @classmethod
+    def tearDownClass(cls) -> None:
+        print(f"Deleting test workspace {cls.workspace_namespace}/{cls.workspace_name}")
+        fapi.delete_workspace(cls.workspace_namespace, cls.workspace_name)
+
+    def tearDown(self) -> None:
+        fapi.delete_entities_of_type(self.workspace_namespace, self.workspace_name, "sample_set")
+        fapi.delete_entities_of_type(self.workspace_namespace, self.workspace_name, "sample")
+        pass
+
+    def run_test(self, samples_tsv=None, expected_sample_sets_json=None, sample_sets_to_mark_delivered=None,
+                 samples_to_mark_rework=None):
+        if sample_sets_to_mark_delivered is None:
+            sample_sets_to_mark_delivered = []
+        if samples_to_mark_rework is None:
+            samples_to_mark_rework = []
+        if samples_tsv is not None:
+            fapi.upload_entities_tsv(self.workspace_namespace, self.workspace_name, samples_tsv, "flexible")
```

Comment:
In theory you're running the risk of uploading a table with the same name as one that already exists and it would retain rows from the old table, which could lead to unexpected behavior during the tests. You could alleviate that by checking if a table with that name already exists and deleting it if necessary.

I'm actually not sure why it works in practice, because the tables do seem to have the same name, but it does, so I could be wrong. Does `subTest` initialize a new class, and therefore a new workspace?

Reply:
I actually want the subtests within a single test to add onto the previous tests, because I am often testing the sequence of runs and sample additions as opposed to just a single run on its own.  This is perhaps not the cleanest way to do this, but I would otherwise need to upload sample_set tables in addition to sample tables to get the workspace into the state I want for the later subtests.

For the separate test methods, `tearDown` clears away the sample_set and sample tables.